Adjust Groovy coverage for Java 20

alllex opened the pull request:
Fixes Groovy integration tests, because Groovy 3.0.13 does not support running with Java 20.

Fixes: https://github.com/gradle/gradle-private/issues/3729

jvandort commented on subprojects/language-groovy/src/testFixtures/groovy/org/gradle/testing/fixture/GroovyCoverage.groovy:
```diff
@@ -46,17 +46,17 @@ class GroovyCoverage {
-        SUPPORTS_PARAMETERS = versionsAbove(SUPPORTED_BY_JDK, "2.5.0")
-        SUPPORTS_DISABLING_AST_TRANSFORMATIONS = versionsAbove(SUPPORTED_BY_JDK, "2.0.0")
-        SINCE_3_0 = versionsAbove(SUPPORTED_BY_JDK, "3.0.0")
+        SUPPORTS_INDY = versionsNotLessThan(SUPPORTED_BY_JDK, "2.3.0")
```
Have you verified that all of these versions are still correct given the change in behavior between `versionsAbove` and `versionsNotLessThan`?

alllex replied:
My changes do not change the semantics. `versionsAbove` was an inclusive version.

jvandort replied:
Shouldn't `versionsNotLessThan ` and `versionsNotGreaterThan` be calling `filterVersionsStrictly` in order to not change behavior, since `filterVersions` behavior has changed

jvandort commented on subprojects/language-groovy/src/testFixtures/groovy/org/gradle/testing/fixture/GroovyCoverage.groovy:
```diff
@@ -99,28 +101,37 @@ class GroovyCoverage {
     }
 
-    private static Set<String> versionsAbove(Collection<String> versionsToFilter, String threshold) {
+    private static Set<String> versionsNotLessThan(Collection<String> versionsToFilter, String threshold) {
```
`versionsNotLessThan` is a little more difficult to reason about compared to something like `versionsAtLeast` or `versionsGreaterThanOrEqual`. Thoughts?

alllex commented on subprojects/language-groovy/src/testFixtures/groovy/org/gradle/testing/fixture/GroovyCoverage.groovy:
```diff
@@ -99,28 +101,37 @@ class GroovyCoverage {
+    }
+
     private static Set<String> versionsBetween(Collection<String> versionsToFilter, String lowerBound, String upperBound) {
         filterVersions(versionsToFilter, lowerBound, upperBound)
```
Me too, this is why I renamed most of the methods!

jvandort replied:
But this currently calls `filterVersions` which is the inclusive version

alllex commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"6e4f9ad6a7ea35601e039218c26d5a2a537b93d2","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you:
- [PullRequestFeedback build](https://builds.gradle.org/viewQueued.html?itemId=62136924)

alllex commented:
@jvandort you mention in the comments that I changed the semantics of version selection, but I don’t think I did. At least, it was not my intention.

The confusion may come from the fact that all methods in the `GroovyCoverage` _before_ the changes were inclusive. So `versionsAbove`, `versionsBelow` and `versionsBetween` were all inclusive.
In the PR, I effectively renamed those to more explicit versions. But I agree that the naming I chose was not the easiest to read.

I pushed another commit polishing those names:
- `versionsAbove` (old inclusive semantics) -> `versionsAtLeast`
- `versionsBelow` -> `versionsAtMost`
- `versionsBetween` -> `versionsBetweenInclusive`

I also added another method `versionsAbove` that now has an **exclusive** semantics.

I hope this answers most of the comments.

alllex commented on subprojects/language-groovy/src/testFixtures/groovy/org/gradle/testing/fixture/GroovyCoverage.groovy:
```diff
@@ -46,16 +46,16 @@ class GroovyCoverage {
             ? GroovySystem.version
-            : versionsBelow(SUPPORTED_BY_JDK, GroovySystem.version).last()
+            : versionsAtMost(SUPPORTED_BY_JDK, GroovySystem.version).last()
         MINIMAL_GROOVY_3 = versionsBelow(SINCE_3_0, "4.0.0").first()
```
This was previously a bug, because `versionsBelow` was inclusive, so potentially it could have returned `4.0.0`. After these changes, `versionsBelow` is exclusive.

alllex commented:
@bot-gradle test and merge

bot-gradle commented:
I've triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=62872981) for you.
